Remove pinning from number formatting in corelib

stephentoub opened the pull request:
The primary focus was on removing places where we were pinning strings and spans.  In some places then for consistency, I switched some things that were using stack-allocated pointers to use references, even though doing so doesn't actually remove any pinning.  These changes also allowed removing some workarounds for span-based stackalloc lifetime semantics.  There were some minor perf differences, both improvements and regressions, but they all looked to be within the noise margins.

cc: @jkotas, @tannergooding

jkotas commented:
My personal preference is to use pointers where everything is equal. Pointers make the code more succinct and easier to read. Short-term pinning has negligible cost.

stephentoub commented:
> My personal preference is to use pointers where everything is equal. Pointers make the code more succinct and easier to read. Short-term pinning has negligible cost.

Meaning you'd prefer we just close this?

jkotas commented on src/libraries/System.Private.CoreLib/src/System/Number.Formatting.cs:
```diff
@@ -532,16 +526,15 @@ private static int GetFloatingPointMaxDigitsAndPrecision(char fmt, ref int preci
             }
 
-            NumberBuffer number = new NumberBuffer(NumberBufferKind.FloatingPoint, pDigits, DoubleNumberBufferLength);
+            byte* pDigits = stackalloc byte[DoubleNumberBufferLength];
+            NumberBuffer number = new NumberBuffer(NumberBufferKind.FloatingPoint, new Span<byte>(pDigits, DoubleNumberBufferLength));
```
```suggestion
            NumberBuffer number = new NumberBuffer(NumberBufferKind.FloatingPoint, stackalloc byte[DoubleNumberBufferLength]);
```
Is there a reason why this has the pDigits temp local? The local was removed in other places.

stephentoub replied:
This and two other places couldn't remove it due to C# ref lifetime rules (the builder is passed in as an argument and then both the builder and this are passed by ref to another method).

jkotas commented:
> Meaning you'd prefer we just close this?

I do not have a strong opinion on this. I know that there has been an excitement about using byrefs and `Unsafe.*` in libraries (both and 3rd party) where regular unmanaged pointers would work just fine.

ericstj commented on src/libraries/Common/src/System/Text/ValueStringBuilder.cs:
```diff
@@ -243,7 +246,8 @@ public unsafe void Append(char* value, int length)
             {
-                dst[i] = *value++;
+                dst[i] = value;
+                value = ref Unsafe.Add(ref value, 1);
```
Looks like you'll need a PackageReference in a place where you now use this https://dev.azure.com/dnceng/public/_build/results?buildId=1636207&view=logs&j=81a0f8c2-d4db-561f-9b8e-3f045280d7fe&t=2a104984-c43b-5b93-1361-6bd8a1e96b04&l=1991

stephentoub committed:
Remove pinning from number formatting in corelib

The primary focus was on removing places where we were pinning strings and spans.  In some places then for consistency, I switched some things that were using stack-allocated pointers to use references, even though doing so doesn't actually remove any pinning.  These changes also allowed removing some workarounds for span-based stackalloc lifetime semantics.

stephentoub commented:
I'm going to close this.  I spent some more time looking at benchmarks, and it's mostly in the noise.  Default integer formatting with ToString() does appear to be up to 10% faster on my machine, but that's the only thing that seems to be measurably better, and if it makes maintenance harder across the board, it's not worth it.
